lib/matplotlib/axes/_axes.py

```diff
@@ -6091,30 +6091,36 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
                              "Please only use 'density', since 'normed'"
                              "will be deprecated.")
 
-        # process the unit information
-        self._process_unit_info(xdata=x, kwargs=kwargs)
-        x = self.convert_xunits(x)
-        if bin_range is not None:
-            bin_range = self.convert_xunits(bin_range)
-
-        # Check whether bins or range are given explicitly.
-        binsgiven = (cbook.iterable(bins) or bin_range is not None)
-
         # basic input validation
         input_empty = np.size(x) == 0
-
         # Massage 'x' for processing.
         if input_empty:
             x = np.array([[]])
         else:
             x = cbook._reshape_2D(x, 'x')
         nx = len(x)  # number of datasets
 
+        # Process unit information
+        # If doing a stacked histogram, the input is a list of datasets, so
+        # we need to do the unit conversion individually on each dataset
+        if stacked:
+            self._process_unit_info(xdata=x[0], kwargs=kwargs)
+            x = [self.convert_xunits(xi) for xi in x]
```

Comment:
I don't understand: it seems to me that the criterion for needing this step is not whether the histogram is stacked, but whether `x` is a list of datasets.

Reply:
I think you're right, I'll change this and add a test for multiple datasets that aren't stacked